### PR TITLE
runtime/syntax: Highlight syntax inside f-string replacement fields in Python

### PR DESCRIPTION
**Pull request by f4nb0y**
Highlight syntax inside [f-string] replacement fields

Attn: @zvezdan
Refs: #10734, #14033

[f-string]: https://docs.python.org/3/reference/lexical_analysis.html#f-strings

### CODE REVIEW

**Comment by f4nb0y** [comment]
This PR is unfinished: Lines 143-145, 147, 149, 151, 153, 162 and 163 have premature delimiter boundaries.

image

Adding `pythonStatement` and `pythonNumber` to the exclusion list fixes some (but not all) of these regressions, while disabling expected highlighting.

**Comment by zvezdan** [comment]
It's encouraging that the same approach with `contains` that was already used in the syntax file for some other matches works in most of the cases here too. Thanks for looking into it!

**Comment by f4nb0y** [comment]
From the [docs]:

> When "keepend" is not used, a match with an end pattern is retried after each contained match.

This is expected. What is less expected is that a contained match also appears to override skip patterns. In the errors above, the colon is end matched after a contained statement or number, even though the contained match is surrounded by a skip match.

To resolve this, we move the skip pattern to a contained match. This may have side-effects on other uses of `contains=ALLBUT` in this file, but we do not have tests to verify this.

[docs]: https://vimhelp.org/syntax.txt.html#%3Asyn-keepend

**Comment by f4nb0y** [comment]
Hi @zvezdan, does this change cause issues with `pythonMatrixMultiply` or `pythonDoctest`?

**Comment by dkearns** [comment]
Thanks for your work.

I noticed a few things in passing:

- `{{` in `f"foo{{bar"` should be matched as pythonEscape or something more specific
- expressions are not matched in format specifiers - `width + 42` is not matched as an expression in `f"result: {value:{width + 42}}"`
- `\N{` is improperly causing the replacement field to be escaped in `Fr'Raw f-string with literal \u00A1 and \U00010605 and \N{INVERTED EXCLAMATION MARK} and fields {foo} and {bar}'`, `INVERTED EXCLAMATION MARK` is a bad expression here.

I confess I'm not quite sure what motivated you to reimplement much of #14057 but it looks like these examples are properly handled by that one already.  If you're planning on matching that PR I can close it and maybe make later PRs to address some of the edge cases.  I've kept it up because, although unfinished, I know it's in use in the wild.

**Comment by zvezdan** [comment]
> Hi @zvezdan, does this change cause issues with `pythonMatrixMultiply` or `pythonDoctest`?

I don't see any issues. Matrix multiplication looks correct in both regular code and doctests.
Similarly, all variations of the strings look good in doctests.
Thanks a lot @f4nb0y!

**Comment by zvezdan** [comment]
> * `{{` in `f"foo{{bar"` should be matched as pythonEscape or something more specific

It is colored the same as the rest of the string, which is the exact intent of `{{` and `}}`. It's _not_ a pair of `{` and `}` that are colored differently and start a field inside the string. The double braces are visually distinct already and keeping them the same color as the string shows they are a part of it.

> * expressions are not matched in format specifiers - `width + 42` is not matched as an expression in `f"result: {value:{width + 42}}"`

They show in format specifier color, which keeps them distinct from a possible expression in front of that format. No need to confuse field expression and the format expression. A single color for the whole format seems fine to me.

> * `\N{` is improperly causing the replacement field to be escaped in `Fr'Raw f-string with literal \u00A1 and \U00010605 and \N{INVERTED EXCLAMATION MARK} and fields {foo} and {bar}'`, `INVERTED EXCLAMATION MARK` is a bad expression here.

I don't see that. I just tried, in both regular code and doctests:
```
v = fr'\u00A1 \U00010605 \N{INVERTED EXCLAMATION MARK} {spam} {eggs}'
```
The whole f-string is colored the same, including both `\U` and `\N` sequences, except `{spam}` and `{eggs}`.
IOW, the "escapes" are not honored, because it's a raw string and they are rendered as regular string characters.
Only the fields are rendered differently, as expected.
Perhaps you're not using the latest version of the syntax file + only this diff, and perhaps you have something else applied?

**Review by zvezdan — APPROVED** [review]
@chrisbra Let's merge this please.
@f4nb0y Thanks!

**Comment by f4nb0y** [comment]
@dkearns said:

> * `\N{` is improperly causing the replacement field to be escaped in `Fr'Raw f-string with literal \u00A1 and \U00010605 and \N{INVERTED EXCLAMATION MARK} and fields {foo} and {bar}'`, `INVERTED EXCLAMATION MARK` is a bad expression here.

Oh dear. You are totally correct:

    % python3 python_strings_bytes.py
      File "python_strings_bytes.py", line 83
        test = Fr'Raw f-string with literal \u00A1 and \U00010605 and \N{INVERTED EXCLAMATION MARK} and fields {foo} and {bar}'
                                                                         ^^^^^^^^^^^^^^^^^^^^
    SyntaxError: invalid syntax. Perhaps you forgot a comma?

I was so busy checking the syntax that I never tested the actual code :-)

**Comment by f4nb0y** [comment]
Additionally, Unicode escapes in lines 168, 173 and 176 raise `SyntaxWarning`, which is reasonable but should be noted.

**Comment by dkearns** [comment]
> > * `{{` in `f"foo{{bar"` should be matched as pythonEscape or something more specific
> 
> It is colored the same as the rest of the string, which is the exact intent of `{{` and `}}`. It's _not_ a pair of `{` and `}` that are colored differently and start a field inside the string. The double braces are visually distinct already and keeping them the same color as the string shows they are a part of it.

It's an escape sequence not two literal `{` characters, highlighting it the same as the rest of the string hides this fact. It's the same as `\\` in `"foo\\bar"`. Likewise, the `}}`.
 
> > * expressions are not matched in format specifiers - `width + 42` is not matched as an expression in `f"result: {value:{width + 42}}"`
> 
> They show in format specifier color, which keeps them distinct from a possible expression in front of that format. No need to confuse field expression and the format expression. A single color for the whole format seems fine to me.

OK,  but I don't really understand why some expressions are worthy of being highlighted and other's not.  The format specifiers can be quite complex and the more specific the highlighting the better.

> > * `\N{` is improperly causing the replacement field to be escaped in `Fr'Raw f-string with literal \u00A1 and \U00010605 and \N{INVERTED EXCLAMATION MARK} and fields {foo} and {bar}'`, `INVERTED EXCLAMATION MARK` is a bad expression here.
> 
> I don't see that. I just tried, in both regular code and doctests:
> 
> ```
> v = fr'\u00A1 \U00010605 \N{INVERTED EXCLAMATION MARK} {spam} {eggs}'
> ```
> 
> The whole f-string is colored the same, including both `\U` and `\N` sequences, except `{spam}` and `{eggs}`. IOW, the "escapes" are not honored, because it's a raw string and they are rendered as regular string characters. Only the fields are rendered differently, as expected. Perhaps you're not using the latest version of the syntax file + only this diff, and perhaps you have something else applied?

That's the bug.  `{INVERTED EXCLAMATION MARK}` is a replacement field.  As escape sequences are not respected in raw strings the `\N{` has no special meaning and is ignored, however, the `{` still introduces a replacement field.

```python
print(fr'\u00A1 \U00010605 \N{INVERTED EXCLAMATION MARK} {spam} {eggs}')
```

```
File "/tmp/foo.py", line 7
    (INVERTED EXCLAMATION MARK)
     ^^^^^^^^^^^^^^^^^^^^
SyntaxError: f-string: invalid syntax. Perhaps you forgot a comma?

shell returned 1
```

File: runtime/syntax/python.vim [hunk 1]
```diff
@@ -211,18 +211,19 @@ syn region  pythonRawBytes
...
 "     - Matched braces inside a format specification are ignored
 "
 syn region  pythonFStringField
     \ matchgroup=pythonFStringDelimiter
     \ start=/{/
-    \ skip=/([^)]*)\|\[[^]]*]\|{[^}]*}\|#.*$/
-    \ end=/\%(\s*=\s*\)\=\%(!\a\)\=\%(:\%({[^}]*}\|[^}]*\)\+\)\=}/
+    \ end=/\%(=\s*\)\=\%(!\a\)\=\%(:\%({[^}]*}\|[^}]*\)\+\)\=}/
+    \ contained
+    \ contains=ALLBUT,pythonFStringField,pythonClass,pythonFunction,pythonDoctest,pythonDoctestValue,@Spell
```
Comment (dkearns) [inline runtime/syntax/python.vim:232]: You're probably aware of this, but this contains value should exclude more syntax groups.  E.g.,

`f"...{NOTE * Warning}..."` will highlight `NOTE` and `Warning` as `pythonTodo` and `pythonException` respectively.

This is a wider problem with the `contains=ALLBUT` strategy but shows up here due to the more generous playing field of `pythonFStringField`.

**Comment by dkearns** [comment]
Conversions with trailing whitespace aren't highlighted.  Other similarly unnecessary whitespace is highlighted properly so this probably should be too.

```python
f"...{42!r }..."
f"...{42!r :<42}..."
f"...{42=!r }..."
```

**Comment by chrisbra** [comment]
> @chrisbra Let's merge this please.

Will do once the last bits have been addressed. Thanks all!

**Comment by f4nb0y** [comment]
@dkearns said:

> Conversions with trailing whitespace aren't highlighted. Other similarly unnecessary whitespace is highlighted properly so this probably should be too.

Interesting. This is a `SyntaxError` in earlier versions:

    >>> sys.version
    '3.11.3 (v3.11.3:f3909b8bc8, Apr  4 2023, 20:12:10) [Clang 13.0.0 (clang-1300.0.29.30)]'
    >>> f'This is a {"string"!a}'
    "This is a 'string'"
    >>> f'This is a {"string"!a }'
      File "<stdin>", line 1
        f'This is a {"string"!a }'
                                  ^
    SyntaxError: f-string: expecting '}'

But it works fine in 3.13. So we should support it.

File: runtime/syntax/python.vim (hunk 1 above)
Comment (f4nb0y) [reply runtime/syntax/python.vim:232]: Hmmm. In f-strings and doctests, we only want top level syntax. I wonder if `contains=TOP` would be more reliable than `contains=ALLBUT` ?

Let's open a separate issue for this.

**Comment by zvezdan** [comment]
> > > * `{{` in `f"foo{{bar"` should be matched as pythonEscape or something more specific
> > 
> > It is colored the same as the rest of the string, which is the exact intent of `{{` and `}}`. It's _not_ a pair of `{` and `}` that are colored differently and start a field inside the string. The double braces are visually distinct already and keeping them the same color as the string shows they are a part of it.
> 
> It's an escape sequence not two literal `{` characters, highlighting it the same as the rest of the string hides this fact. It's the same as `\\` in `"foo\\bar"`. Likewise, the `}}`.
> 
> > > * expressions are not matched in format specifiers - `width + 42` is not matched as an expression in `f"result: {value:{width + 42}}"`
> > 
> > 
> > They show in format specifier color, which keeps them distinct from a possible expression in front of that format. No need to confuse field expression and the format expression. A single color for the whole format seems fine to me.
> 
> OK, but I don't really understand why some expressions are worthy of being highlighted and other's not. The format specifiers can be quite complex and the more specific the highlighting the better.
> 

In general, I agree ...
My thinking, for both of these topics, was that this PR does not need to be blocked on the above requirements.
Incremental improvements are OK and I feel that the above can be added later.
The PR feels fine and complete as is, adding a nice incremental improvements/fixes to previous PRs on the same topic.

> > > * `\N{` is improperly causing the replacement field to be escaped in `Fr'Raw f-string with literal \u00A1 and \U00010605 and \N{INVERTED EXCLAMATION MARK} and fields {foo} and {bar}'`, `INVERTED EXCLAMATION MARK` is a bad expression here.
> > 
> > I don't see that. I just tried, in both regular code and doctests:
> > ```
> > v = fr'\u00A1 \U00010605 \N{INVERTED EXCLAMATION MARK} {spam} {eggs}'
> > ```
> >   
> > The whole f-string is colored the same, including both `\U` and `\N` sequences, except `{spam}` and `{eggs}`. IOW, the "escapes" are not honored, because it's a raw string and they are rendered as regular string characters. Only the fields are rendered differently, as expected. Perhaps you're not using the latest version of the syntax file + only this diff, and perhaps you have something else applied?
> 
> That's the bug. `{INVERTED EXCLAMATION MARK}` is a replacement field. As escape sequences are not respected in raw strings the `\N{` has no special meaning and is ignored, however, the `{` still introduces a replacement field.
> 
> ```python
> print(fr'\u00A1 \U00010605 \N{INVERTED EXCLAMATION MARK} {spam} {eggs}')
> ```
> 
>  ```
> File "/tmp/foo.py", line 7
>     (INVERTED EXCLAMATION MARK)
>      ^^^^^^^^^^^^^^^^^^^^
> SyntaxError: f-string: invalid syntax. Perhaps you forgot a comma?
> 
> shell returned 1
> ```

OK. This was a misunderstanding. I thought you were saying that the color in syntax highlight was wrong. You were referring to an actual syntax error and possibly the need to indicate that somehow in the highlighting. Perhaps another incremental improvement candidate?

**Comment by dkearns** [comment]
> In general, I agree ... My thinking, for both of these topics, was that this PR does not need to be blocked on the above requirements. Incremental improvements are OK and I feel that the above can be added later. The PR feels fine and complete as is, adding a nice incremental improvements/fixes to previous PRs on the same topic.

Sure, I'm not trying to block anything being merged I'm just responding to @f4nb0y's [request for comments](https://github.com/vim/vim/issues/14033#issuecomment-3193032727).

I have some more issues but they're not likely to crop up in basic usage so I'll leave them until after this is merged.

> OK. This was a misunderstanding. I thought you were saying that the color in syntax highlight was wrong. You were referring to an actual syntax error and possibly the need to indicate that somehow in the highlighting. Perhaps another incremental improvement candidate?

No, the highlighting was wrong but it has since been addressed by https://github.com/vim/vim/pull/17962/commits/0d1cfe2c6d1f63ee0d99b09c68389bb3aa49c1bd

**Comment by dkearns** [comment]
> @dkearns said:
> 
> > Conversions with trailing whitespace aren't highlighted. Other similarly unnecessary whitespace is highlighted properly so this probably should be too.
> 
> Interesting. This is a `SyntaxError` in earlier versions:
> 
> ```
> >>> sys.version
> '3.11.3 (v3.11.3:f3909b8bc8, Apr  4 2023, 20:12:10) [Clang 13.0.0 (clang-1300.0.29.30)]'
> >>> f'This is a {"string"!a}'
> "This is a 'string'"
> >>> f'This is a {"string"!a }'
>   File "<stdin>", line 1
>     f'This is a {"string"!a }'
>                               ^
> SyntaxError: f-string: expecting '}'
> ```
> 
> But it works fine in 3.13. So we should support it.

The grammar has been relaxed in a number of areas over time.

**Comment by dkearns** [comment]
I'm a bit busy elsewhere and haven't looked at the implementation yet but it seems nested replacement fields are often highlighted incorrectly.  In these examples, the last `}` is matched as a string character as the replacement field is ended prematurely by a nested expression.

```python
f'{42:<{x}}'
f'{42:{x}.{y}}'
```

**Comment by f4nb0y** [comment]
@dkearns said:

> I'm a bit busy elsewhere and haven't looked at the implementation yet but it seems nested replacement fields are often highlighted incorrectly.

Indeed they are. Looking at it now.

**Comment by dkearns** [comment]
Floating point numbers aren't matched at the start of replacement  fields `f"{42.01}"`, this is matched as `42`.

It looks like this is due to the float pattern starting with `\%(^\|\W\)\zs\d` rather than `\<\d` to account for the optional integer part.  That second float pattern could be broken up into multiple patterns or maybe just replace the `\zs` with a width-constrained lookbehind.

**Comment by f4nb0y** [comment]
@dkearns said:

> I'm a bit busy elsewhere and haven't looked at the implementation yet but it seems nested replacement fields are often highlighted incorrectly.

Is this better?

**Comment by Felixoid** [comment]
Testing it for a few days, no issues so far

**Comment by f4nb0y** [comment]
@dkearns said:

> Floating point numbers aren't matched at the start of replacement fields `f"{42.01}"`, this is matched as `42`.

True, but I'm not updating `pythonNumber` in this PR. That needs regression tests that we don't have.

**Comment by f4nb0y** [comment]
Hi @dkearns and @zvezdan,

Are there any other edge cases to address before we merge this?

**Comment by f4nb0y** [comment]
Hi @chrisbra, I think this is ready to merge?

**Comment by chrisbra** [comment]
thanks all, let me merge it now

**Comment by zeertzjq** [comment]
This seems to highlight all variable names in replacement fields as `pythonType`, since `pythonType` isn't specified after `ALLBUT,`. Is this intended?

**Comment by f4nb0y** [comment]
@zeertzjq said:

> This seems to highlight all variable names in replacement fields as pythonType, since pythonType isn't specified after ALLBUT,. Is this intended?

It is not. Opened #18206 to address this.
